cloud/scope/vpc_cluster.go

```diff
@@ -166,3 +197,281 @@ func (s *VPCClusterScope) Close() error {
+	return s.GetResourceGroupID()
+}
+
+// GetResourceGroupID returns the Resource Group ID for the cluster.
+func (s *VPCClusterScope) GetResourceGroupID() (string, error) {
+	// Check if the Resource Group ID is available from Status first.
+	if s.IBMVPCCluster.Status.ResourceGroup != nil && s.IBMVPCCluster.Status.ResourceGroup.ID != "" {
+		return s.IBMVPCCluster.Status.ResourceGroup.ID, nil
+	}
+
+	// If the Resource Group is not defined in Spec, we generate the name based on the cluster name.
+	resourceGroupName := s.IBMVPCCluster.Spec.ResourceGroup
+	if resourceGroupName == "" {
+		resourceGroupName = s.IBMVPCCluster.Name
+	}
+
+	// Retrieve the Resource Group based on the name.
+	resourceGroup, err := s.ResourceManagerClient.GetResourceGroupByName(resourceGroupName)
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve resource group by name: %w", err)
+	} else if resourceGroup == nil || resourceGroup.ID == nil {
+		return "", fmt.Errorf("failed to find resource group by name: %s", resourceGroupName)
+	}
+
+	// Populate the Stauts Resource Group to shortcut future lookups.
+	s.SetResourceStatus(infrav1beta2.ResourceTypeResourceGroup, &infrav1beta2.ResourceStatus{
+		ID:    *resourceGroup.ID,
+		Name:  ptr.To(resourceGroupName),
+		Ready: true,
+	})
+
+	return *resourceGroup.ID, nil
+}
+
+// GetServiceName returns the name of a given service type from Spec or generates a name for it.
+func (s *VPCClusterScope) GetServiceName(resourceType infrav1beta2.ResourceType) *string {
+	switch resourceType {
+	case infrav1beta2.ResourceTypeVPC:
+		// Generate a name based off cluster name if no VPC defined in Spec, or no VPC name nor ID.
+		if s.NetworkSpec().VPC == nil || (s.NetworkSpec().VPC.Name == nil && s.NetworkSpec().VPC.ID == nil) {
+			return ptr.To(fmt.Sprintf("%s-vpc", s.Name()))
+		}
+		if s.NetworkSpec().VPC.Name != nil {
+			return s.NetworkSpec().VPC.Name
+		}
+	default:
+		s.V(3).Info("unsupported resource type", "resourceType", resourceType)
+	}
+	return nil
+}
+
+// GetVPCID returns the VPC id, if available.
+func (s *VPCClusterScope) GetVPCID() (*string, error) {
+	// Check if the VPC ID is available from Status first.
+	if s.NetworkStatus() != nil && s.NetworkStatus().VPC != nil {
+		return ptr.To(s.NetworkStatus().VPC.ID), nil
+	}
+
+	if s.NetworkSpec() != nil && s.NetworkSpec().VPC != nil {
+		if s.NetworkSpec().VPC.ID != nil {
+			return s.NetworkSpec().VPC.ID, nil
+		} else if s.NetworkSpec().VPC.Name != nil {
+			vpcDetails, err := s.VPCClient.GetVPCByName(*s.NetworkSpec().VPC.Name)
+			if err != nil {
+				return nil, fmt.Errorf("failed vpc id lookup: %w", err)
+			}
+
+			// Check if the VPC was found and has an ID
+			if vpcDetails != nil && vpcDetails.ID != nil {
+				// Set VPC ID in Status to shortcut future lookups
+				s.SetResourceStatus(infrav1beta2.ResourceTypeVPC, &infrav1beta2.ResourceStatus{
+					ID:    *vpcDetails.ID,
+					Name:  s.NetworkSpec().VPC.Name,
+					Ready: false,
```

Comment:
Any reasons for setting Ready to false here?

Reply:
Updated